File: python/semantic_kernel/connectors/ai/function_choice_behavior.py
# Copyright (c) Microsoft. All rights reserved.

import logging
from collections.abc import Callable
from enum import Enum
from typing import TYPE_CHECKING, Literal, TypeVar

from typing_extensions import deprecated

from semantic_kernel.connectors.ai.function_calling_utils import _combine_filter_dicts
from semantic_kernel.exceptions.service_exceptions import ServiceInitializationError
from semantic_kernel.kernel_pydantic import KernelBaseModel
from semantic_kernel.utils.experimental_decorator import experimental_class

if TYPE_CHECKING:
    from semantic_kernel.connectors.ai.function_call_behavior import FunctionCallBehavior
    from semantic_kernel.connectors.ai.function_call_choice_configuration import FunctionCallChoiceConfiguration
    from semantic_kernel.connectors.ai.prompt_execution_settings import PromptExecutionSettings
    from semantic_kernel.kernel import Kernel


DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS = 5

logger = logging.getLogger(__name__)


_T = TypeVar("_T", bound="FunctionChoiceBehavior")


@experimental_class
class FunctionChoiceType(Enum):
    """The type of function choice behavior."""

    AUTO = "auto"
    NONE = "none"
    REQUIRED = "required"


@experimental_class
class FunctionChoiceBehavior(KernelBaseModel):
    """Class that controls function choice behavior.

    Attributes:
        enable_kernel_functions: Enable kernel functions.
        max_auto_invoke_attempts: The maximum number of auto invoke attempts.
        filters: Filters for the function choice behavior. Available options are: excluded_plugins,
            included_plugins, excluded_functions, or included_functions.
        type_: The type of function choice behavior.

    Properties:
        auto_invoke_kernel_functions: Check if the kernel functions should be auto-invoked.
            Determined as max_auto_invoke_attempts > 0.

    Methods:
        configure: Configures the settings for the function call behavior,
            the default version in this class, does nothing, use subclasses for different behaviors.

    Class methods:
        Auto: Returns FunctionChoiceBehavior class with auto_invoke enabled, and the desired functions
            based on either the specified filters or the full qualified names. The model will decide which function
            to use, if any.
        NoneInvoke: Returns FunctionChoiceBehavior class with auto_invoke disabled, and the desired functions
            based on either the specified filters or the full qualified names. The model does not invoke any functions,
            but can rather describe how it would invoke a function to complete a given task/query.
        Required: Returns FunctionChoiceBehavior class with auto_invoke enabled, and the desired functions
            based on either the specified filters or the full qualified names. The model is required to use one of the
            provided functions to complete a given task/query.
    """

    enable_kernel_functions: bool = True
    maximum_auto_invoke_attempts: int = DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS
    filters: (
        dict[Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]]
        | None
    ) = None
    type_: FunctionChoiceType | None = None

    @classmethod
    @deprecated("The `FunctionCallBehavior` class is deprecated; use `FunctionChoiceBehavior` instead.")
    def from_function_call_behavior(cls: type[_T], behavior: "FunctionCallBehavior") -> _T:
        """Create a FunctionChoiceBehavior from a FunctionCallBehavior."""
        from semantic_kernel.connectors.ai.function_call_behavior import (
            EnabledFunctions,
            KernelFunctions,
            RequiredFunction,
        )

        if isinstance(behavior, (EnabledFunctions, KernelFunctions)):
            return cls.Auto(
                auto_invoke=behavior.auto_invoke_kernel_functions,
                filters=behavior.filters if hasattr(behavior, "filters") else None,
            )
        if isinstance(behavior, (RequiredFunction)):
            return cls.Required(
                auto_invoke=behavior.auto_invoke_kernel_functions,
                filters={"included_functions": [behavior.function_fully_qualified_name]},
            )
        return cls(
            enable_kernel_functions=behavior.enable_kernel_functions,
            maximum_auto_invoke_attempts=behavior.max_auto_invoke_attempts,
        )

    @property
    def auto_invoke_kernel_functions(self):
        """Return True if auto_invoke_kernel_functions is enabled."""
        return self.maximum_auto_invoke_attempts > 0

    @auto_invoke_kernel_functions.setter
    def auto_invoke_kernel_functions(self, value: bool):
        """Set the auto_invoke_kernel_functions property."""
        self.maximum_auto_invoke_attempts = DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS if value else 0

    def _check_and_get_config(
        self,
        kernel: "Kernel",
        filters: dict[
            Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]
        ]
        | None = {},
    ) -> "FunctionCallChoiceConfiguration":
        """Check for missing functions and get the function call choice configuration."""
        from semantic_kernel.connectors.ai.function_call_choice_configuration import FunctionCallChoiceConfiguration

        if filters:
            return FunctionCallChoiceConfiguration(available_functions=kernel.get_list_of_function_metadata(filters))
        return FunctionCallChoiceConfiguration(available_functions=kernel.get_full_list_of_function_metadata())

    def configure(
        self,
        kernel: "Kernel",
        update_settings_callback: Callable[..., None],
        settings: "PromptExecutionSettings",
    ) -> None:
        """Configure the function choice behavior."""
        if not self.enable_kernel_functions:
            return

        config = self.get_config(kernel)

        if config:
            update_settings_callback(config, settings, self.type_)

    def get_config(self, kernel: "Kernel") -> "FunctionCallChoiceConfiguration":
        """Get the function call choice configuration based on the type."""
        return self._check_and_get_config(kernel, self.filters)

    @classmethod
    def Auto(
        cls: type[_T],
        auto_invoke: bool = True,
        *,
        filters: dict[
            Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]
        ]
        | None = None,
        **kwargs,
    ) -> _T:
        """Creates a FunctionChoiceBehavior with type AUTO.

        Returns FunctionChoiceBehavior class with auto_invoke enabled, and the desired functions
        based on either the specified filters or the full qualified names. The model will decide which function
        to use, if any.
        """
        kwargs.setdefault("maximum_auto_invoke_attempts", DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS if auto_invoke else 0)
        return cls(
            type_=FunctionChoiceType.AUTO,
            filters=filters,
            **kwargs,
        )

    @classmethod
    def NoneInvoke(
        cls: type[_T],
        *,
        filters: dict[
            Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]
        ]
        | None = None,
        **kwargs,
    ) -> _T:
        """Creates a FunctionChoiceBehavior with type NONE.

        Returns FunctionChoiceBehavior class with auto_invoke disabled, and the desired functions
        based on either the specified filters or the full qualified names. The model does not invoke any functions,
        but can rather describe how it would invoke a function to complete a given task/query.
        """
        kwargs.setdefault("maximum_auto_invoke_attempts", 0)
        return cls(
            type_=FunctionChoiceType.NONE,
            filters=filters,
            **kwargs,
        )

    @classmethod
    def Required(
        cls: type[_T],
        auto_invoke: bool = True,
        *,
        filters: dict[
            Literal["excluded_plugins", "included_plugins", "excluded_functions", "included_functions"], list[str]
        ]
        | None = None,
        **kwargs,
    ) -> _T:
        """Creates a FunctionChoiceBehavior with type REQUIRED.

        Returns FunctionChoiceBehavior class with auto_invoke enabled, and the desired functions
        based on either the specified filters or the full qualified names. The model is required to use one of the
        provided functions to complete a given task/query.
        """
        kwargs.setdefault("maximum_auto_invoke_attempts", 1 if auto_invoke else 0)
        return cls(
            type_=FunctionChoiceType.REQUIRED,
            filters=filters,
            **kwargs,
        )

    @classmethod
    def from_dict(cls: type[_T], data: dict) -> _T:
        """Create a FunctionChoiceBehavior from a dictionary."""
        type_map = {
            "auto": cls.Auto,
            "none": cls.NoneInvoke,
            "required": cls.Required,
        }
        behavior_type = data.pop("type", "auto")
        auto_invoke = data.pop("auto_invoke", False)
        functions = data.pop("functions", None)
        filters = data.pop("filters", None)

        if functions:
            valid_fqns = [name.replace(".", "-") for name in functions]
            if filters:
                filters = _combine_filter_dicts(filters, {"included_functions": valid_fqns})
            else:
                filters = {"included_functions": valid_fqns}

        return type_map[behavior_type](  # type: ignore
            auto_invoke=auto_invoke,
            filters=filters,
            **data,
        )

    @classmethod
    def from_string(cls: type[_T], data: str) -> _T:
        """Create a FunctionChoiceBehavior from a string.

        This method converts the provided string to a FunctionChoiceBehavior object
        for the specified type.
        """
        type_value = data.lower()
        if type_value == "auto":
            return cls.Auto()
        if type_value == "none":
            return cls.NoneInvoke()
        if type_value == "required":
            return cls.Required()
        raise ServiceInitializationError(
            f"The specified type `{type_value}` is not supported. Allowed types are: `auto`, `none`, `required`."
        )

File: nodejs/semantic-kernel/connectors/ai/function-choice-behavior.ts
// Copyright (c) Microsoft. All rights reserved.

import { FunctionCallChoiceConfiguration } from './function-call-choice-configuration'
import { combineFilterDicts } from './function-calling-utils'
import { FunctionChoiceType } from './function-choice-type'

const DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS = 5

/**
 * Filter types for function choice behavior.
 */
export type FunctionFilters = {
  excludedPlugins?: string[]
  includedPlugins?: string[]
  excludedFunctions?: string[]
  includedFunctions?: string[]
}

/**
 * Prompt execution settings interface (minimal).
 */
export interface PromptExecutionSettings {
  [key: string]: any
}

/**
 * Kernel interface (minimal).
 */
export interface Kernel {
  getListOfFunctionMetadata(filters: FunctionFilters): any[]
  getFullListOfFunctionMetadata(): any[]
}

/**
 * Update settings callback type.
 */
export type UpdateSettingsCallback = (
  config: FunctionCallChoiceConfiguration,
  settings: PromptExecutionSettings,
  type?: FunctionChoiceType
) => void

/**
 * Class that controls function choice behavior.
 *
 * @experimental This class is experimental and may change in the future.
 */
export class FunctionChoiceBehavior {
  enableKernelFunctions: boolean = true
  maximumAutoInvokeAttempts: number = DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS
  filters?: FunctionFilters
  type?: FunctionChoiceType

  constructor(options?: {
    enableKernelFunctions?: boolean
    maximumAutoInvokeAttempts?: number
    filters?: FunctionFilters
    type?: FunctionChoiceType
  }) {
    if (options) {
      this.enableKernelFunctions = options.enableKernelFunctions ?? true
      this.maximumAutoInvokeAttempts = options.maximumAutoInvokeAttempts ?? DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS
      this.filters = options.filters
      this.type = options.type
    }
  }

  /**
   * Check if the kernel functions should be auto-invoked.
   * Determined as maximumAutoInvokeAttempts > 0.
   */
  get autoInvokeKernelFunctions(): boolean {
    return this.maximumAutoInvokeAttempts > 0
  }

  /**
   * Set the auto_invoke_kernel_functions property.
   */
  set autoInvokeKernelFunctions(value: boolean) {
    this.maximumAutoInvokeAttempts = value ? DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS : 0
  }

  /**
   * Check for missing functions and get the function call choice configuration.
   *
   * @private
   */
  private checkAndGetConfig(kernel: Kernel, filters?: FunctionFilters): FunctionCallChoiceConfiguration {
    if (filters) {
      return new FunctionCallChoiceConfiguration({
        availableFunctions: kernel.getListOfFunctionMetadata(filters),
      })
    }
    return new FunctionCallChoiceConfiguration({
      availableFunctions: kernel.getFullListOfFunctionMetadata(),
    })
  }

  /**
   * Configure the function choice behavior.
   *
   * @param kernel - The kernel instance
   * @param updateSettingsCallback - Callback to update settings
   * @param settings - The prompt execution settings
   */
  configure(kernel: Kernel, updateSettingsCallback: UpdateSettingsCallback, settings: PromptExecutionSettings): void {
    if (!this.enableKernelFunctions) {
      return
    }

    const config = this.getConfig(kernel)

    if (config) {
      updateSettingsCallback(config, settings, this.type)
    }
  }

  /**
   * Get the function call choice configuration based on the type.
   *
   * @param kernel - The kernel instance
   * @returns The function call choice configuration
   */
  getConfig(kernel: Kernel): FunctionCallChoiceConfiguration {
    return this.checkAndGetConfig(kernel, this.filters)
  }

  /**
   * Creates a FunctionChoiceBehavior with type AUTO.
   *
   * Returns FunctionChoiceBehavior with auto_invoke enabled, and the desired functions
   * based on either the specified filters or the full qualified names. The model will decide which function
   * to use, if any.
   *
   * @param autoInvoke - Whether to enable auto invoke (default: true)
   * @param filters - Filters for function selection
   * @param options - Additional options
   * @returns A new FunctionChoiceBehavior instance
   */
  static Auto(
    autoInvoke: boolean = true,
    filters?: FunctionFilters,
    options?: Partial<FunctionChoiceBehavior>
  ): FunctionChoiceBehavior {
    const maximumAutoInvokeAttempts =
      options?.maximumAutoInvokeAttempts ?? (autoInvoke ? DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS : 0)

    return new FunctionChoiceBehavior({
      type: FunctionChoiceType.AUTO,
      filters,
      maximumAutoInvokeAttempts,
      ...options,
    })
  }

  /**
   * Creates a FunctionChoiceBehavior with type NONE.
   *
   * Returns FunctionChoiceBehavior with auto_invoke disabled, and the desired functions
   * based on either the specified filters or the full qualified names. The model does not invoke any functions,
   * but can rather describe how it would invoke a function to complete a given task/query.
   *
   * @param filters - Filters for function selection
   * @param options - Additional options
   * @returns A new FunctionChoiceBehavior instance
   */
  static NoneInvoke(filters?: FunctionFilters, options?: Partial<FunctionChoiceBehavior>): FunctionChoiceBehavior {
    const maximumAutoInvokeAttempts = options?.maximumAutoInvokeAttempts ?? 0

    return new FunctionChoiceBehavior({
      type: FunctionChoiceType.NONE,
      filters,
      maximumAutoInvokeAttempts,
      ...options,
    })
  }

  /**
   * Creates a FunctionChoiceBehavior with type REQUIRED.
   *
   * Returns FunctionChoiceBehavior with auto_invoke enabled, and the desired functions
   * based on either the specified filters or the full qualified names. The model is required to use one of the
   * provided functions to complete a given task/query.
   *
   * @param autoInvoke - Whether to enable auto invoke (default: true)
   * @param filters - Filters for function selection
   * @param options - Additional options
   * @returns A new FunctionChoiceBehavior instance
   */
  static Required(
    autoInvoke: boolean = true,
    filters?: FunctionFilters,
    options?: Partial<FunctionChoiceBehavior>
  ): FunctionChoiceBehavior {
    const maximumAutoInvokeAttempts = options?.maximumAutoInvokeAttempts ?? (autoInvoke ? 1 : 0)

    return new FunctionChoiceBehavior({
      type: FunctionChoiceType.REQUIRED,
      filters,
      maximumAutoInvokeAttempts,
      ...options,
    })
  }

  /**
   * Create a FunctionChoiceBehavior from a dictionary.
   *
   * @param data - The data object
   * @returns A new FunctionChoiceBehavior instance
   */
  static fromDict(data: Record<string, any>): FunctionChoiceBehavior {
    const dataCopy = { ...data }
    const behaviorType = dataCopy.type || 'auto'
    const autoInvoke = dataCopy.autoInvoke || false
    const functions = dataCopy.functions
    let filters = dataCopy.filters

    delete dataCopy.type
    delete dataCopy.autoInvoke
    delete dataCopy.functions
    delete dataCopy.filters

    if (functions && Array.isArray(functions)) {
      const validFqns = functions.map((name: string) => name.replace('.', '-'))
      if (filters) {
        // Convert FunctionFilters to Record<string, string[]> for combineFilterDicts
        const filtersDict: Record<string, string[]> = {}
        if (filters.excludedPlugins) filtersDict.excludedPlugins = filters.excludedPlugins
        if (filters.includedPlugins) filtersDict.includedPlugins = filters.includedPlugins
        if (filters.excludedFunctions) filtersDict.excludedFunctions = filters.excludedFunctions
        if (filters.includedFunctions) filtersDict.includedFunctions = filters.includedFunctions

        const combined = combineFilterDicts(filtersDict, { includedFunctions: validFqns })
        filters = combined as FunctionFilters
      } else {
        filters = { includedFunctions: validFqns }
      }
    }

    const typeMap: Record<
      string,
      (autoInvoke: boolean, filters?: FunctionFilters, options?: any) => FunctionChoiceBehavior
    > = {
      auto: (ai, f, o) => FunctionChoiceBehavior.Auto(ai, f, o),
      none: (_ai, f, o) => FunctionChoiceBehavior.NoneInvoke(f, o),
      required: (ai, f, o) => FunctionChoiceBehavior.Required(ai, f, o),
    }

    const creator = typeMap[behaviorType.toLowerCase()]
    if (!creator) {
      throw new Error(`Unknown function choice behavior type: ${behaviorType}`)
    }

    return creator(autoInvoke, filters, dataCopy)
  }

  /**
   * Create a FunctionChoiceBehavior from a string.
   *
   * This method converts the provided string to a FunctionChoiceBehavior object
   * for the specified type.
   *
   * @param data - The type string ('auto', 'none', or 'required')
   * @returns A new FunctionChoiceBehavior instance
   */
  static fromString(data: string): FunctionChoiceBehavior {
    const typeValue = data.toLowerCase()

    if (typeValue === 'auto') {
      return FunctionChoiceBehavior.Auto()
    }
    if (typeValue === 'none') {
      return FunctionChoiceBehavior.NoneInvoke()
    }
    if (typeValue === 'required') {
      return FunctionChoiceBehavior.Required()
    }

    throw new Error(
      `The specified type \`${typeValue}\` is not supported. Allowed types are: \`auto\`, \`none\`, \`required\`.`
    )
  }
}
